app/test/testMetadata.js

```diff
@@ -0,0 +1,161 @@
+const chai = require('chai');
+const chaiHttp = require('chai-http');
+const fetchMock = require('fetch-mock');
+const {app} = require('../service');
+
+chai.use(chaiHttp);
+chai.should();
+
+const exampleMetadata = {
+  description: 'An example description',
+  favicon_url: 'http://www.example.com/rich-icon.png',
+  icon_url: 'http://www.example.com/rich-icon.png',
+  image_url: 'http://www.example.com/preview.png',
+  images: [
+    {
+      'url': 'http://www.example.com/preview.png',
+      'entropy': 1,
+      'height': 500,
+      'width': 500,
+    }
+  ],
+  original_url: 'http://www.example.com/good',
+  provider_url: 'http://www.example.com/good',
+  title: 'An Example Page',
+  url: 'http://www.example.com/good'
+};
+
+const goodExampleUrl = 'http://www.example.com/good';
+const badExampleUrl = 'http://www.example.com/bad';
+
+
+describe('Metadata API Tests', function() {
+  before(function() {
+    fetchMock.mock(
+      goodExampleUrl,
+      `
+        <html>
+          <head>
+            <title>${exampleMetadata.title}</title>
+            <meta name="description" content="${exampleMetadata.description}" />
+            <meta name="thumbnail" content="${exampleMetadata.images[0].url}" />
+            <link rel="icon" href="http://www.example.com/rich-icon.png" />
+          </head>
+          <body></body>
+        </html>
+      `
+    );
+    fetchMock.mock(badExampleUrl, 500);
+  });
+
+  after(function() {
+    fetchMock.restore();
+  });
+
+  it('should raise 400 if missing JSON body', (done) => {
+    chai.request(app)
+      .post('/')
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should raise 400 if missing urls parameter', (done) => {
+    chai.request(app)
+      .post('/')
+      .send(JSON.stringify({}))
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should raise 400 if urls is empty', (done) => {
+    chai.request(app)
+      .post('/')
+      .send(JSON.stringify({urls: []}))
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should raise 400 if content header is not set', (done) => {
+    chai.request(app)
+      .post('/')
+      .send(JSON.stringify({urls: [goodExampleUrl]}))
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should return 200 with metadata for a url', (done) => {
+    chai.request(app)
+      .post('/')
+      .set('content-type', 'application/json')
+      .send(JSON.stringify({urls: [goodExampleUrl]}))
+      .end((err, res) => {
+
+        res.should.have.status(200);
+
+        const expectedResponse = {
+          error: '',
+          urls: {}
+        };
+        expectedResponse.urls[goodExampleUrl] = exampleMetadata;
+
+        res.body.should.deep.equal(expectedResponse);
+
+        done();
+      });
+  });
```

Comment:
Another Promise vs callback example (which uses `.then()` block instead of the `.catch()`):

```js
  it('should return 200 with metadata for a url', () => {
    return chai.request(app)
      .post('/')
      .set('content-type', 'application/json')
      .send(JSON.stringify({urls: [goodExampleUrl]}))
      .then((res) => {
        res.should.have.status(200);

        const expectedResponse = {
          error: '',
          urls: {}
        };

        expectedResponse.urls[goodExampleUrl] = exampleMetadata;
        res.body.should.deep.equal(expectedResponse);
      });
  });
```